packages/router-api/src/primitives.rs

```diff
@@ -245,6 +223,219 @@ impl AsRef<str> for ChainName {
+        String::from_utf8(value)
+            .map_err(StdError::invalid_utf8)?
+            .then(LegacyChainName::try_from)
+            .map_err(StdError::invalid_utf8)
+    }
+}
+
+impl KeyDeserialize for &LegacyChainName {
+    type Output = ChainName;
+
+    #[inline(always)]
+    fn from_vec(value: Vec<u8>) -> StdResult<Self::Output> {
+        ChainName::from_vec(value)
+    }
+}
+
+#[cw_serde]
+#[serde(try_from = "String", into = "String")]
+#[derive(Eq, Hash)]
+pub enum GeneralizedChainName {
+    Amplifier(ChainName),
+    Legacy(LegacyChainName),
+}
+
+impl GeneralizedChainName {
+    pub fn amplifier(self) -> Result<ChainName, Error> {
+        match self {
+            GeneralizedChainName::Amplifier(chain) => Ok(chain),
+            GeneralizedChainName::Legacy(_) => Err(Error::InvalidChainName),
+        }
+    }
+
+    pub fn legacy(self) -> Result<LegacyChainName, Error> {
+        match self {
+            GeneralizedChainName::Amplifier(_) => Err(Error::InvalidChainName),
+            GeneralizedChainName::Legacy(chain) => Ok(chain),
+        }
+    }
+}
+
+impl AsRef<str> for GeneralizedChainName {
+    fn as_ref(&self) -> &str {
+        match self {
+            GeneralizedChainName::Amplifier(chain) => chain.as_ref(),
+            GeneralizedChainName::Legacy(chain) => chain.as_ref(),
+        }
+    }
+}
+
+impl FromStr for GeneralizedChainName {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        ChainName::from_str(s).map_or_else(
+            |_| Ok(GeneralizedChainName::Legacy(LegacyChainName::from_str(s)?)),
+            |chain| Ok(GeneralizedChainName::Amplifier(chain)),
+        )
+    }
+}
+
+impl TryFrom<String> for GeneralizedChainName {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl From<GeneralizedChainName> for String {
+    fn from(d: GeneralizedChainName) -> Self {
+        match d {
+            GeneralizedChainName::Amplifier(chain) => chain.into(),
+            GeneralizedChainName::Legacy(chain) => chain.into(),
+        }
+    }
+}
+
+impl Display for GeneralizedChainName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            GeneralizedChainName::Amplifier(chain) => Display::fmt(chain, f),
+            GeneralizedChainName::Legacy(chain) => Display::fmt(chain, f),
+        }
+    }
+}
+
+impl PartialEq<ChainName> for GeneralizedChainName {
+    fn eq(&self, other: &ChainName) -> bool {
+        match self {
+            GeneralizedChainName::Amplifier(chain) => chain.eq(other),
+            GeneralizedChainName::Legacy(_) => false,
+        }
+    }
+}
+
+impl PartialEq<LegacyChainName> for GeneralizedChainName {
+    fn eq(&self, other: &LegacyChainName) -> bool {
+        match self {
+            GeneralizedChainName::Amplifier(_) => false,
+            GeneralizedChainName::Legacy(chain) => chain.eq(other),
+        }
+    }
+}
```

Comment:
Just curious, why do we need to define this? It looks like a default implementation to me.